Warnings: expression result unused; implicit conversion changes signedness

clang-5 prints a lot of warnings, like these:
```
In file included from staged-include/xtensor/xarray.hpp:19:
staged-include/xtensor/xcontainer.hpp:1235:20: warning: expression result unused [-Wunused-value]
            (void)(c, size);  // remove unused parameter warning
                   ^

staged-include/xtensor/xslice.hpp:354:41: warning: implicit conversion changes signedness: 'std::size_t' (aka 'unsigned long') to 'xt::xall<long>::size_type' (aka 'long') [-Wsign-conversion]
            return xall<std::ptrdiff_t>(size);
                   ~~~~                 ^~~~
```

--- a/include/xtensor/xbuilder.hpp
+++ b/include/xtensor/xbuilder.hpp
@@ -478,7 +478,7 @@ namespace xt
             template <class It>
             inline value_type element(It first, It) const
             {
-                return m_source(*(first + m_axis));
+                return m_source(*(first + static_cast<std::ptrdiff_t>(m_axis)));
             }
 
         private:

--- a/include/xtensor/xcontainer.hpp
+++ b/include/xtensor/xcontainer.hpp
@@ -1232,7 +1232,8 @@ namespace xt
         template <class C, class S>
         inline void resize_data_container(const C& c, S size)
         {
-            (void)(c, size);  // remove unused parameter warning
+            (void) c;  // remove unused parameter warning
+            (void) size;
             XTENSOR_ASSERT_MSG(c.size() == size, "Trying to resize const data container with wrong size.");
         }
     }

--- a/include/xtensor/xexception.hpp
+++ b/include/xtensor/xexception.hpp
@@ -127,12 +127,15 @@ namespace xt
     inline void check_element_index(const S& shape, It first, It last)
     {
         auto dst = static_cast<typename S::size_type>(last - first);
-        It efirst = last - std::min(shape.size(), dst);
+        It efirst = last - static_cast<std::ptrdiff_t>(std::min(shape.size(), dst));
         std::size_t axis = 0;
         while (efirst != last)
         {
+        #pragma GCC diagnostic push
+        #pragma GCC diagnostic ignored "-Wsign-compare"
             if (*efirst >= shape[axis] && shape[axis] != 1)
             {
+        #pragma GCC diagnostic pop
                 throw std::out_of_range("index " + std::to_string(*efirst) + " is out of bounds for axis "
                     + std::to_string(axis) + " with size " + std::to_string(shape[axis]));
             }

--- a/include/xtensor/xiterator.hpp
+++ b/include/xtensor/xiterator.hpp
@@ -1098,7 +1098,8 @@ namespace xt
     template <class It, class BIt>
     inline auto xbounded_iterator<It, BIt>::operator*() const -> value_type
     {
-        return (*m_it < *m_bound_it) ? *m_it : static_cast<value_type>((*m_bound_it) - 1);
+        using type = decltype(*m_bound_it);
+        return (static_cast<type>(*m_it) < *m_bound_it) ? *m_it : static_cast<value_type>((*m_bound_it) - 1);
     }
 
     template <class It, class BIt>

--- a/include/xtensor/xmath.hpp
+++ b/include/xtensor/xmath.hpp
@@ -303,6 +303,9 @@ INT_SPECIALIZATION_IMPL(FUNC_NAME, RETURN_VAL, unsigned long long);             
 
         // The following specializations are needed to avoid 'ambiguous overload' errors,
         // whereas 'unsigned char' and 'unsigned short' are automatically converted to 'int'.
+        // we're still adding those functions to silence warnings
+        UNSIGNED_ABS_FUNC(unsigned char);
+        UNSIGNED_ABS_FUNC(unsigned short);
         UNSIGNED_ABS_FUNC(unsigned int);
         UNSIGNED_ABS_FUNC(unsigned long);
         UNSIGNED_ABS_FUNC(unsigned long long);

--- a/include/xtensor/xslice.hpp
+++ b/include/xtensor/xslice.hpp
@@ -259,8 +259,8 @@ namespace xt
 
         template <class MI = A, class MA = B, class STEP = C>
         inline std::enable_if_t<std::is_integral<MI>::value &&
-                                    std::is_integral<MA>::value &&
-                                    std::is_integral<STEP>::value,
+                                std::is_integral<MA>::value &&
+                                std::is_integral<STEP>::value,
                                 xstepped_range<std::ptrdiff_t>>
             get(std::size_t /*size*/) const
         {
@@ -269,28 +269,28 @@ namespace xt
 
         template <class MI = A, class MA = B, class STEP = C>
         inline std::enable_if_t<!std::is_integral<MI>::value &&
-                                    std::is_integral<MA>::value &&
-                                    std::is_integral<STEP>::value,
+                                std::is_integral<MA>::value &&
+                                std::is_integral<STEP>::value,
                                 xstepped_range<std::ptrdiff_t>>
         get(std::size_t size) const
         {
-            return xstepped_range<std::ptrdiff_t>(m_step > 0 ? 0 : int(size) - 1, m_max, m_step);
+            return xstepped_range<std::ptrdiff_t>(m_step > 0 ? 0 : static_cast<std::ptrdiff_t>(size) - 1, m_max, m_step);
         }
 
         template <class MI = A, class MA = B, class STEP = C>
         inline std::enable_if_t<std::is_integral<MI>::value &&
-                                    !std::is_integral<MA>::value &&
-                                    std::is_integral<STEP>::value,
+                                !std::is_integral<MA>::value &&
+                                std::is_integral<STEP>::value,
                                 xstepped_range<std::ptrdiff_t>>
         get(std::size_t size) const
         {
-            return xstepped_range<std::ptrdiff_t>(m_min, m_step > 0 ? int(size) : -1, m_step);
+            return xstepped_range<std::ptrdiff_t>(m_min, m_step > 0 ? static_cast<std::ptrdiff_t>(size) : -1, m_step);
         }
 
         template <class MI = A, class MA = B, class STEP = C>
         inline std::enable_if_t<std::is_integral<MI>::value &&
-                                    std::is_integral<MA>::value &&
-                                    !std::is_integral<STEP>::value,
+                                std::is_integral<MA>::value &&
+                                !std::is_integral<STEP>::value,
                                 xrange<std::ptrdiff_t>>
             get(std::size_t /*size*/) const
         {
@@ -299,44 +299,44 @@ namespace xt
 
         template <class MI = A, class MA = B, class STEP = C>
         inline std::enable_if_t<!std::is_integral<MI>::value &&
-                                    !std::is_integral<MA>::value &&
-                                    std::is_integral<STEP>::value,
+                                !std::is_integral<MA>::value &&
+                                std::is_integral<STEP>::value,
                                 xstepped_range<std::ptrdiff_t>>
         get(std::size_t size) const
         {
-            int min_val_arg = m_step > 0 ? 0 : int(size) - 1;
-            int max_val_arg = m_step > 0 ? int(size) : -1;
+            std::ptrdiff_t min_val_arg = m_step > 0 ? 0 : static_cast<std::ptrdiff_t>(size) - 1;
+            std::ptrdiff_t max_val_arg = m_step > 0 ? static_cast<std::ptrdiff_t>(size) : -1;
             return xstepped_range<std::ptrdiff_t>(min_val_arg, max_val_arg, m_step);
         }
 
         template <class MI = A, class MA = B, class STEP = C>
         inline std::enable_if_t<std::is_integral<MI>::value &&
-                                    !std::is_integral<MA>::value &&
-                                    !std::is_integral<STEP>::value,
+                                !std::is_integral<MA>::value &&
+                                !std::is_integral<STEP>::value,
                                 xrange<std::ptrdiff_t>>
         get(std::size_t size) const
         {
-            return xrange<std::ptrdiff_t>(std::size_t(m_min), size);
+            return xrange<std::ptrdiff_t>(static_cast<std::ptrdiff_t>(m_min), static_cast<std::ptrdiff_t>(size));
         }
 
         template <class MI = A, class MA = B, class STEP = C>
         inline std::enable_if_t<!std::is_integral<MI>::value &&
-                                    std::is_integral<MA>::value &&
-                                    !std::is_integral<STEP>::value,
+                                std::is_integral<MA>::value &&
+                                !std::is_integral<STEP>::value,
                                 xrange<std::ptrdiff_t>>
             get(std::size_t /*size*/) const
         {
-            return xrange<std::ptrdiff_t>(0, std::size_t(m_max));
+            return xrange<std::ptrdiff_t>(0, static_cast<std::ptrdiff_t>(m_max));
         }
 
         template <class MI = A, class MA = B, class STEP = C>
         inline std::enable_if_t<!std::is_integral<MI>::value &&
-                                    !std::is_integral<MA>::value &&
-                                    !std::is_integral<STEP>::value,
+                                !std::is_integral<MA>::value &&
+                                !std::is_integral<STEP>::value,
                                 xall<std::ptrdiff_t>>
         get(std::size_t size) const
         {
-            return xall<std::ptrdiff_t>(size);
+            return xall<std::ptrdiff_t>(static_cast<std::ptrdiff_t>(size));
         }
 
     private:

--- a/include/xtensor/xstrided_view.hpp
+++ b/include/xtensor/xstrided_view.hpp
@@ -726,7 +726,7 @@ namespace xt
     template <class It>
     inline It xstrided_view<CT, S, FS>::data_xbegin_impl(It begin) const noexcept
     {
-        return begin + m_offset;
+        return begin + static_cast<std::ptrdiff_t>(m_offset);
     }
 
     template <class CT, class S, class FS>

--- a/test/test_xbroadcast.cpp
+++ b/test/test_xbroadcast.cpp
@@ -69,8 +69,9 @@ namespace xt
 
     TEST(xbroadcast, shape_forwarding)
     {
-        std::array<std::size_t, 2> bc_shape;
+        std::array<std::size_t, 2> bc_shape{3, 3};
         auto m1_broadcast = broadcast(123, bc_shape);
+        ASSERT_EQ(m1_broadcast(0, 0), 123);
     }
 
     TEST(xbroadcast, xiterator)

--- a/test/test_xnan_functions.cpp
+++ b/test/test_xnan_functions.cpp
@@ -77,9 +77,9 @@ namespace xt
         EXPECT_EQ(nancumsum(nantest::aN), cumsum(nantest::aR));
         EXPECT_EQ(nancumsum(nantest::aN, 0), cumsum(nantest::aR, 0));
         EXPECT_EQ(nancumsum(nantest::aN, 1), cumsum(nantest::aR, 1));
-        EXPECT_EQ(nancumsum(nantest::xN, {0}), cumsum(nantest::xR, {0}));
-        EXPECT_EQ(nancumsum(nantest::xN, {1}), cumsum(nantest::xR, {1}));
-        EXPECT_EQ(nancumsum(nantest::xN, {2}), cumsum(nantest::xR, {2}));
+        EXPECT_EQ(nancumsum(nantest::xN, 0), cumsum(nantest::xR, 0));
+        EXPECT_EQ(nancumsum(nantest::xN, 1), cumsum(nantest::xR, 1));
+        EXPECT_EQ(nancumsum(nantest::xN, 2), cumsum(nantest::xR, 2));
     }
 
     TEST(xnanfunctions, multid)
@@ -90,11 +90,11 @@ namespace xt
         dynamic_view(arr, {0, xt::ellipsis()}) = nanv;
         dynamic_view(carr, {0, xt::ellipsis()}) = 0;
 
-        EXPECT_EQ(nancumsum(arr, {0}), cumsum(carr, {0}));
-        EXPECT_EQ(nancumsum(arr, {1}), cumsum(carr, {1}));
-        EXPECT_EQ(nancumsum(arr, {2}), cumsum(carr, {2}));
-        EXPECT_EQ(nancumsum(arr, {3}), cumsum(carr, {3}));
-        EXPECT_EQ(nancumsum(arr, {4}), cumsum(carr, {4}));
+        EXPECT_EQ(nancumsum(arr, 0), cumsum(carr, 0));
+        EXPECT_EQ(nancumsum(arr, 1), cumsum(carr, 1));
+        EXPECT_EQ(nancumsum(arr, 2), cumsum(carr, 2));
+        EXPECT_EQ(nancumsum(arr, 3), cumsum(carr, 3));
+        EXPECT_EQ(nancumsum(arr, 4), cumsum(carr, 4));
     }
 
     TEST(xnanfunctions, nancumprod)
@@ -102,9 +102,9 @@ namespace xt
         EXPECT_EQ(nancumprod(nantest::aN), cumprod(nantest::aP));
         EXPECT_EQ(nancumprod(nantest::aN, 0), cumprod(nantest::aP, 0));
         EXPECT_EQ(nancumprod(nantest::aN, 1), cumprod(nantest::aP, 1));
-        EXPECT_EQ(nancumprod(nantest::xN, {0}), cumprod(nantest::xP, {0}));
-        EXPECT_EQ(nancumprod(nantest::xN, {1}), cumprod(nantest::xP, {1}));
-        EXPECT_EQ(nancumprod(nantest::xN, {2}), cumprod(nantest::xP, {2}));
+        EXPECT_EQ(nancumprod(nantest::xN, 0), cumprod(nantest::xP, 0));
+        EXPECT_EQ(nancumprod(nantest::xN, 1), cumprod(nantest::xP, 1));
+        EXPECT_EQ(nancumprod(nantest::xN, 2), cumprod(nantest::xP, 2));
     }
 
 }

